upstairs/src/downstairs.rs

```diff
@@ -2286,57 +2148,76 @@ impl Downstairs {
         }
     }
 
-    /// Enqueue a [DownstairsIO] job:
+    /// Enqueue a [IOop] job:
     ///
     /// - enqueue the job in each of [Self::clients] (clients may skip the job)
     /// - add the job to [Self::ds_active]
     /// - Mark the job as ackable if it was skipped by all downstairs
     /// - Check that the job was already acked if it's a write (the "fast ack"
     ///   optimization, which is performed elsewhere)
     /// - Send the job to each downstairs client task (if not skipped)
-    fn enqueue(&mut self, mut io: DownstairsIO) {
+    fn enqueue(
+        &mut self,
+        ds_id: JobId,
+        guest_id: GuestWorkId,
+        io: IOop,
+        mut bp_guard: ClientMap<BackpressureGuard>,
+    ) {
         let mut skipped = 0;
         let last_repair_extent = self.last_repair_extent();
 
         // Send the job to each client!
-        let mut needs_send = ClientData::new(false);
-        for cid in ClientId::iter() {
-            let job_state =
-                self.clients[cid].enqueue(&mut io, last_repair_extent);
-            if matches!(job_state, IOState::Skipped) {
-                skipped += 1;
+        let state = ClientData::from_fn(|cid| {
+            let client = &mut self.clients[cid];
+            if client.enqueue(ds_id, &io, last_repair_extent) {
+                // Update the per-client backpressure guard
+                if !bp_guard.contains(&cid) {
+                    let g = client.backpressure_counters.increment(&io);
+                    bp_guard.insert(cid, g);
+                }
+                self.send(ds_id, io.clone(), cid);
+                IOState::InProgress
             } else {
-                assert_eq!(job_state, IOState::New);
-                needs_send[cid] = true;
+                skipped += 1;
+                IOState::Skipped
             }
-        }
+        });
 
-        let is_write = io.work.is_write();
-        let ds_id = io.ds_id;
+        let is_write = io.is_write();
         if skipped == 3 {
             if !is_write {
                 self.ackable_work.insert(ds_id);
             }
             warn!(self.log, "job {} skipped on all downstairs", &ds_id);
         }
-        assert_eq!(is_write, io.acked);
 
         // Puts the IO onto the downstairs work queue.
-        self.ds_active.insert(ds_id, io);
-
-        for cid in ClientId::iter() {
-            if needs_send[cid] {
-                self.send(ds_id, cid);
-            }
-        }
+        self.ds_active.insert(
+            ds_id,
+            DownstairsIO {
+                ds_id,
+                guest_id,
+                work: io,
+                state,
+                acked: is_write,
+                replay: false,
+                data: None,
+                read_validations: vec![],
+                backpressure_guard: bp_guard,
+            },
+        );
     }
 
     /// Marks the given job as in-progress for this client, then sends it
```

Comment:
The job arrives at `send` without a per-client state, because it's (now) a `(JobId, IOop)` tuple; `state` is a part of the `DownstairsIO`, which has not yet been constructed.  I updated the comment to remove the "in-progress" reference